engine/class_modules/paladin/sc_paladin.hpp

```diff
@@ -1082,18 +1081,168 @@ struct sanctified_wrath_t : public paladin_spell_t
   }
 };
 
-
-struct holy_power_consumer_t : public paladin_melee_attack_t
+template <class Base >
+struct holy_power_consumer_t : public Base
 {
+  private:
+    typedef Base ab; // action base, eg. spell_t
+  public:
+    typedef holy_power_consumer_t base_t;
   bool is_divine_storm;
-  holy_power_consumer_t( const std::string& n, paladin_t* p, const spell_data_t* s ) :
-    paladin_melee_attack_t( n, p, s ),
+  holy_power_consumer_t( const std::string& n, paladin_t* player,
+      const spell_data_t* s = spell_data_t::nil() ) :
```

Comment:
I don't think defaulting spelldata to nil is a great idea, that should only be there for auto attacks.